lib/workers/repository/update/pr/code-owners.ts

```diff
@@ -21,37 +32,131 @@ export async function codeOwnersForPr(pr: Pr): Promise<string[]> {
-    const rules = codeOwnersFile
+
+    if (!prFiles?.length) {
+      logger.debug('PR includes no files');
+      return [];
+    }
+
+    // Convert CODEOWNERS file into list of matching rules
+    const fileOwnerRules = codeOwnersFile
       .split(newlineRegex)
+      // Remove empty and commented lines
       .map((line) => line.trim())
       .filter((line) => line && !line.startsWith('#'))
+
+      // Extract pattern & usernames
       .map((line) => {
         const [pattern, ...usernames] = line.split(regEx(/\s+/));
+        const matchPattern = ignore().add(pattern);
         return {
           usernames,
-          match: (path: string) => {
-            const matcher = ignore().add(pattern);
-            return matcher.ignores(path);
-          },
+          pattern,
+          score: pattern.length,
+          match: (path: string) => matchPattern.ignores(path),
         };
-      })
-      .reverse();
+      });
+
     logger.debug(
-      { prFiles, rules },
+      { prFiles, fileOwnerRules },
       'PR files and rules to match for CODEOWNERS'
     );
-    const matchingRule = rules.find((rule) => prFiles?.every(rule.match));
-    if (!matchingRule) {
-      logger.debug('No matching CODEOWNERS rule found');
-      return [];
-    }
+
+    // Apply rules & get list of owners for each prFile
+    const emptyRules = fileOwnerRules.filter(
+      (rule) => rule.usernames.length === 0
+    );
+    const fileOwners =
+      // Map through all prFiles and match said file(s) with all the rules
+      prFiles
+        .map<{
+          file: string;
+          usernames: UserScore[];
+        }>((file) => {
+          const usersWithScore = fileOwnerRules
+            .map((rule) =>
+              rule.match(file)
+                ? { score: rule.score, usernames: rule.usernames }
+                : null
+            )
+            .reduce<UserScore[]>((acc, match) => {
+              if (!match) {
+                return acc;
+              }
+              for (const user of match.usernames) {
+                if (!acc.find((a) => a.username === user)) {
+                  acc.push({ score: match.score, username: user });
+                }
+              }
+
+              return acc;
+            }, []);
+
+          return { file, usernames: usersWithScore };
+        })
+
+        // Match file again but this time only with emptyRules, to ensure that files which have no owner set remain owner-less
+        .map((fileMatch) => {
+          const matchEmpty = emptyRules.find((rule) =>
+            rule.match(fileMatch.file)
+          );
+          if (matchEmpty) {
+            return { ...fileMatch, usernames: [] };
+          }
+          return fileMatch;
+        });
+
+    logger.debug(
+      `CODEOWNERS matched the following files: ${JSON.stringify(fileOwners)}`
+    );
+
+    // Get list of all matched users and the files they own (reverse keys of fileOwners)
+    const usersWithOwnedFiles = fileOwners.reduce<
```

Comment:
same here, extract the callback function

Reply:
Solved in dbf7502a9541316f3e698d95756a8885f60bd144